src/lib.rs

```diff
@@ -259,20 +364,38 @@ impl Conv<f64> for f32 {
         assert_eq!(x, y as f64);
         y
     }
+    #[inline]
+    fn try_conv(x: f64) -> Result<Self, Error> {
+        let y = x as f32;
+        if x == y as f64 {
+            Ok(y)
+        } else {
+            Err(Error::Inexact)
+        }
+    }
 }
 
 macro_rules! impl_via_digits_check {
     ($x:ty: $y:tt) => {
         impl Conv<$x> for $y {
             #[inline]
-            fn conv(x: $x) -> $y {
+            fn conv(x: $x) -> Self {
                 if cfg!(any(debug_assertions, feature = "assert_digits")) {
-                    let src_ty_bits = u32::conv(size_of::<$x>() * 8);
-                    let src_digits = src_ty_bits - (x.leading_zeros() + x.trailing_zeros());
-                    let dst_digits = core::$y::MANTISSA_DIGITS;
-                    assert!(src_digits <= dst_digits);
+                    Self::try_conv(x).expect("inexact int→float conversion")
```

Comment:
Would suggest "inexact int to float conversion" here for friendliness with non-Unicode environments.